### DISCUSSION TITLE
Multiple foreign tables in a single "or" statement

### DISCUSSION THREAD

**Discussion by luketpena** [supabase/supabase#18997] [Feature Requests]
Currently with Supabase JS you can filter by nested joined tables by just just digging down, like this:
`.eq('table1.table2.table3', value)`

You can also filter an or statement by a foreign table like this:
`.or('column1.eq.value1, column2.eq.value2, { foreignTable: 'table2' })`

But if there are multiple foreign tables we want to filter on in a single "or" statement, this is just not possible. This has led us to created a view for every single table we need to do searching on when the values we are searching are spread across multiple joined tables. So if I am getting something on table 1, but want to search by something from table 2 OR table 3, that just isn't possible right now, which feels like a huge oversight.

My suggestion is to either allow the same drill-down syntax from how it can be used in other filter statements (like eq) or to allow us to provide an array of filter statements to an "or" statement where we can specify a different table for each.

My current use case is that I am repeatedly asked by designers to display a bunch of joined data in a table and then have some kind of search bar at the top that is searching by values across multiple joined tables.

**Comment by steve-chavez** [comment]
It is possible to use `or` across tables, but you have to use the `is.null` or `not.is.null` filters. See:

- https://supabase.com/blog/postgrest-11-prerelease#anti-joins
- At the end there's an example with `or` https://postgrest.org/en/stable/references/api/resource_embedding.html#null-filtering-on-embedded-resources

**Comment by rasikhq** [comment]
This is still an issue, found multiple threads on this from multiple people and all of them are just left unanswered because just like OP mentioned, there is no working solution for it besides edge-function/views hell.

@steve-chavez I looked at the links you've posted and couldn't get anything to work still.

Edit: Just moved away from Supabase cuz of this limitation to our own custom backend

**Comment by otang** [comment]
We have a strong need for this as well. My preference for how this would work would be to add support for the table names as prefixes in our filters, similar to how it works with the PostgREST docs. Example:

```
query.or('profiles.first_name.eq.chris,teams.name.eq.Team1')

```

@luketpena did you find any other solution?

@steve-chavez is there any way for us to pull the exit hatch and drop down to a more "raw" level way to query PostgREST through the Supabase client to allow us to create more complex filters across embedded resources?

**Comment by burakkarasel** [comment]
Is there any improvements? I'm facing the same issue.

**Reply by filippobrigati** [reply]
Not yet, but you can create a PostgreSQL view to aggregate data

**Comment by AshkanZiaee** [comment]
Has anyone found a solution to this problem?

**Comment by keneasson** [comment]
I'd propose a callback solution that passes A or B which are both clones of the last function result - if A then use it, OR if B use it.
`
 .eq('t1.id', '123')
 .or((a, b) => return [a.is.('t2.id','not', null), b.gte.('t2.updated_at',timestamp1).lte('t2.update_ad', timestamp2)]
`
Which produces:
WHERE t1.id = '123' AND ((t2.id IS NOT NULL) OR (t2.update_ad > '2024-12-31' AND t2.update_at < 2025-02-01)) ... and you can keep going.

**Comment by NelsonGan** [comment]
We had to do an OR statement on multiple tables (main table and joined table) but it doesn't "just work". I spent hours trying to understand the examples sent by @steve-chavez but it finally works now.

```sql
--- INTENDED SQL QUERY
SELECT 
    table.ID, 
    "Join Table".Amount 
FROM table
JOIN "Join Table" ON table.ID = "Join Table".ID
WHERE 
    --- OR STATEMENT FROM MULTIPLE TABLES
    ("Join Table".Amount < 100) 
    OR ("Join Table".Amount = 100 AND table.ID < 20)
```

```ts
// Does not work
query = query.or(`"Join Table".amount.gt.100, and("Join Table".eq.100, id.gt.20)`)
```

```ts
// Works
let query = supabase.from('Main Table').select(
  `
    id, 
    "Join Table" (
      amount, 
    ),
    deletegateRelation1:"Join Table"(),
    delegateRelation2:"Join Table"()
  `
)

query = query.lt(`deletegateRelation1.amount`, 100)
query = query.eq(`deletegateRelation2.amount`, 100)
query = query.or(`deletegateRelation1.not.is.null, and(deletegateRelation2.not.is.null, id.lt.20)`)
``` 

This is super un-intuitive, but it works. @steve-chavez I think it makes more sense if it support multiple foreign table, do you mind explaining what's the reasoning behind this query choice?

**Reply by laurenceisla** [reply]
> do you mind explaining what's the reasoning behind this query choice?

I can help with the explanation here. The thing is, PostgREST returns the joined tables as a `LEFT JOIN` by default, that's why it doesn't change the top resource in any way. Also, the filters are done at every embedding level, so even if we used `INNER JOIN`, doing an `OR` filter across joins was not possible... until [`null` filtering for embedded tables](https://github.com/PostgREST/postgrest/pull/2584) was introduced. This post https://github.com/PostgREST/postgrest/issues/2563#issuecomment-1336573928 explains the underlying query that PostgREST generates to handle `OR` across joins, it can clarify a bit on the logic used to arrive to this query. There's [an extra example in the docs](https://postgrest.org/en/stable/references/api/resource_embedding.html#or-filtering-across-embedded-resources) that can help too.

> This is super un-intuitive, but it works. @steve-chavez I think it makes more sense if it support multiple foreign table

Yes, agree. We're currently working on making the query more intuitive in this PR: https://github.com/PostgREST/postgrest/pull/4011. Under the hood it would do the same as the SQL statement mentioned above, but the request query will be more user-friendly.

**Reply by rcrogers** [reply]
Thanks for the clear example @NelsonGan !
